File: clients/typescript/src/transports/rest.ts
import { PawrlyError, UnsupportedByTransport } from "../errors.js";
import { QueryHandle, type QueryMeta, type Row } from "../query.js";
import type {
  CacheEntryInfo,
  CatalogSnapshot,
  FunctionDescription,
  FunctionInfo,
  HealthReport,
  MaterializeOutcome,
  MaterializeSpec,
  RefreshCatalogOutcome,
  RefreshOutcome,
  ReloadReport,
  SemanticModelDescription,
  SemanticModelInfo,
  SemanticQuery,
  SourceDef,
  SourceInfo,
  SourceTestReport,
  TableDescription,
  TableInfo,
  VacuumReport,
} from "../result.js";
import type { Transport } from "../transport.js";
import * as convert from "./convert.js";

export class RestTransport implements Transport {
  readonly name = "rest";
  private readonly base: string;

  constructor(
    baseUrl: string,
    private readonly bearer?: string,
  ) {
    this.base = baseUrl.replace(/\/+$/, "");
  }

  private headers(): Headers {
    const h = new Headers();
    h.set("content-type", "application/json");
    if (this.bearer) h.set("authorization", `Bearer ${this.bearer}`);
    return h;
  }

  private async send(path: string, init?: RequestInit): Promise<Record<string, unknown>> {
    const resp = await fetch(`${this.base}${path}`, { ...init, headers: this.headers() });
    const body = (await resp.json().catch(() => ({}))) as Record<string, unknown>;
    if (!resp.ok) {
      const err = (body.error ?? {}) as { code?: string; message?: string };
      throw new PawrlyError(err.code ?? "PAWRLY_INTERNAL", err.message ?? resp.statusText);
    }
    return body;
  }

  async query(
    sql: string,
    params: Record<string, string>,
    limit?: number,
  ): Promise<QueryHandle> {
    // Stream NDJSON so large results stay memory-bounded. NDJSON carries no id
    // or completion envelope, so `id` is empty and `truncated` stays false.
    const payload: Record<string, unknown> = { sql, params, format: "ndjson" };
    if (limit !== undefined) payload.limit = limit;
    const resp = await fetch(`${this.base}/v1/sql`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify(payload),
    });
    if (!resp.ok) {
      const body = (await resp.json().catch(() => ({}))) as Record<string, unknown>;
      const err = (body.error ?? {}) as { code?: string; message?: string };
      throw new PawrlyError(err.code ?? "PAWRLY_INTERNAL", err.message ?? resp.statusText);
    }
    const meta: QueryMeta = { columns: [], rowCount: 0, truncated: false };
    return new QueryHandle("", ndjsonRows(resp.body, meta), meta);
  }

  async semanticQuery(q: SemanticQuery): Promise<QueryHandle> {
    // `/v1/query` returns a buffered JSON envelope (no NDJSON there yet).
    const r = await this.send("/v1/query", {
      method: "POST",
      body: JSON.stringify(semanticToJson(q)),
    });
    return bufferedHandle(
      "",
      (r.columns as string[]) ?? [],
      (r.rows as Row[]) ?? [],
      (r.truncated as boolean) ?? false,
    );
  }

  async explain(sql: string, analyze: boolean): Promise<string> {
    const r = await this.send("/v1/explain", {
      method: "POST",
      body: JSON.stringify({ sql, analyze }),
    });
    return (r.plan as string) ?? "";
  }

  async cancel(queryId: string): Promise<boolean> {
    const r = await this.send(`/v1/queries/${encodeURIComponent(queryId)}/cancel`, {
      method: "POST",
    });
    return (r.cancelled as boolean) ?? false;
  }

  async materialize(
    name: string,
    spec: MaterializeSpec,
    namespace?: string,
  ): Promise<MaterializeOutcome> {
    const r = await this.send(`/v1/materialized/${encodeURIComponent(name)}${nsQuery(namespace)}`, {
      method: "PUT",
      body: JSON.stringify(spec),
    });
    requireNamespaceEcho(namespace, r.namespace as string | undefined);
    const n = (r.name ?? {}) as { schema?: string; table?: string };
    return {
      name: { schema: n.schema ?? "", table: n.table ?? "" },
      filePath: (r.file_path as string) ?? "",
      rowCount: (r.row_count as number) ?? 0,
      sizeBytes: (r.size_bytes as number) ?? 0,
    };
  }

  async listSources(): Promise<SourceInfo[]> {
    const r = await this.send("/v1/sources");
    return ((r.sources as Record<string, unknown>[]) ?? []).map(convert.sourceInfo);
  }

  async listTables(source?: string, nameGlob?: string): Promise<TableInfo[]> {
    const q = new URLSearchParams();
    if (source !== undefined) q.set("source", source);
    if (nameGlob !== undefined) q.set("name_glob", nameGlob);
    const r = await this.send(`/v1/tables${qs(q)}`);
    return ((r.tables as Record<string, unknown>[]) ?? []).map(convert.tableInfo);
  }

  async describeTable(name: string): Promise<TableDescription> {
    return convert.tableDescription(await this.send(`/v1/tables/${encodeURIComponent(name)}`));
  }

  async schemaSnapshot(sources?: string[], compact?: boolean): Promise<CatalogSnapshot> {
    const q = new URLSearchParams();
    if (sources?.length) q.set("sources", sources.join(","));
    if (compact) q.set("compact", "true");
    return convert.catalogSnapshot(await this.send(`/v1/schema${qs(q)}`));
  }

  async cacheEntries(namespace?: string): Promise<CacheEntryInfo[]> {
    const r = await this.send(`/v1/cache${nsQuery(namespace)}`);
    requireNamespaceEcho(namespace, r.namespace as string | undefined);
    return ((r.entries as Record<string, unknown>[]) ?? []).map(convert.cacheEntry);
  }

  async listFunctions(): Promise<FunctionInfo[]> {
    const r = await this.send("/v1/functions");
    return ((r.functions as Record<string, unknown>[]) ?? []).map(convert.functionInfo);
  }

  async describeFunction(namespace: string, name: string): Promise<FunctionDescription> {
    const path = `/v1/functions/${encodeURIComponent(namespace)}/${encodeURIComponent(name)}`;
    return convert.functionDescription(await this.send(path));
  }

  async listSemanticModels(): Promise<SemanticModelInfo[]> {
    const r = await this.send("/v1/semantic/models");
    return ((r.models as Record<string, unknown>[]) ?? []).map(convert.semanticModelInfo);
  }

  async describeSemanticModel(name: string): Promise<SemanticModelDescription> {
    return convert.semanticModelDescription(
      await this.send(`/v1/semantic/models/${encodeURIComponent(name)}`),
    );
  }

  async addSource(def: SourceDef): Promise<SourceInfo> {
    return convert.sourceInfo(
      await this.send("/v1/sources", { method: "POST", body: JSON.stringify(def) }),
    );
  }

  async removeSource(name: string): Promise<boolean> {
    const r = await this.send(`/v1/sources/${encodeURIComponent(name)}`, { method: "DELETE" });
    return (r.removed as boolean) ?? false;
  }

  async testSource(name: string): Promise<SourceTestReport> {
    const path = `/v1/sources/${encodeURIComponent(name)}/test`;
    return convert.sourceTestReport(await this.send(path, { method: "POST" }));
  }

  async reloadConfig(): Promise<ReloadReport> {
    return convert.reloadReport(await this.send("/v1/config/reload", { method: "POST" }));
  }

  async refreshCatalog(source?: string): Promise<RefreshCatalogOutcome> {
    const q = new URLSearchParams();
    if (source !== undefined) q.set("source", source);
    return convert.refreshCatalogOutcome(
      await this.send(`/v1/catalog/refresh${qs(q)}`, { method: "POST" }),
    );
  }

  async refreshTable(name: string): Promise<RefreshOutcome> {
    const path = `/v1/tables/${encodeURIComponent(name)}/refresh`;
    return convert.refreshOutcome(await this.send(path, { method: "POST" }));
  }

  async invalidateCache(name: string): Promise<boolean> {
    const r = await this.send(`/v1/cache/${encodeURIComponent(name)}`, { method: "DELETE" });
    return (r.invalidated as boolean) ?? false;
  }

  async vacuumCache(): Promise<VacuumReport> {
    return convert.vacuumReport(await this.send("/v1/cache/vacuum", { method: "POST" }));
  }

  async dropMaterialized(name: string, namespace?: string): Promise<boolean> {
    const r = await this.send(`/v1/materialized/${encodeURIComponent(name)}${nsQuery(namespace)}`, {
      method: "DELETE",
    });
    requireNamespaceEcho(namespace, r.namespace as string | undefined);
    return (r.dropped as boolean) ?? false;
  }

  async health(): Promise<HealthReport> {
    const r = await this.send("/v1/health");
    return { ok: (r.ok as boolean) ?? false, version: (r.version as string) ?? "" };
  }

  async shutdown(): Promise<void> {
    throw new UnsupportedByTransport("shutdown", "rest");
  }

  close(): void {}
}

function qs(params: URLSearchParams): string {
  const s = params.toString();
  return s ? `?${s}` : "";
}

function nsQuery(namespace?: string): string {
  return namespace ? `?namespace=${encodeURIComponent(namespace)}` : "";
}

function requireNamespaceEcho(requested: string | undefined, echoed: string | undefined): void {
  if (requested && requested !== echoed) {
    throw new PawrlyError(
      "PAWRLY_PROTOCOL",
      `server ignored namespace \`${requested}\` — it predates materialize namespaces, so the ` +
        "operation targeted the default namespace instead; upgrade the server",
    );
  }
}

async function* ndjsonRows(
  body: ReadableStream<Uint8Array> | null,
  meta: QueryMeta,
): AsyncGenerator<Row> {
  if (!body) return;
  const decoder = new TextDecoder();
  let buf = "";
  for await (const chunk of body as unknown as AsyncIterable<Uint8Array>) {
    buf += decoder.decode(chunk, { stream: true });
    let nl: number;
    while ((nl = buf.indexOf("\n")) >= 0) {
      const line = buf.slice(0, nl).trim();
      buf = buf.slice(nl + 1);
      if (line) yield emit(line, meta);
    }
  }
  const tail = (buf + decoder.decode()).trim();
  if (tail) yield emit(tail, meta);
}

function emit(line: string, meta: QueryMeta): Row {
  const row = JSON.parse(line) as Row;
  if (meta.columns.length === 0) meta.columns = Object.keys(row);
  meta.rowCount += 1;
  return row;
}

function bufferedHandle(
  id: string,
  columns: string[],
  rows: Row[],
  truncated: boolean,
): QueryHandle {
  const meta: QueryMeta = { columns, rowCount: rows.length, truncated };
  async function* gen(): AsyncGenerator<Row> {
    for (const r of rows) yield r;
  }
  return new QueryHandle(id, gen(), meta);
}

function semanticToJson(q: SemanticQuery): Record<string, unknown> {
  const body: Record<string, unknown> = {
    measures: q.measures ?? [],
    dimensions: q.dimensions ?? [],
    params: q.params ?? {},
  };
  if (q.filters) {
    body.filters = q.filters.map((f) => ({
      member: f.member,
      op: f.op,
      values: f.values ?? [],
    }));
  }
  if (q.orderBy) {
    body.order_by = q.orderBy.map((o) => ({
      member: o.member,
      direction: o.desc ? "desc" : "asc",
    }));
  }
  if (q.segments) body.segments = q.segments;
  if (q.limit !== undefined) body.limit = q.limit;
  if (q.timeZone) body.time_zone = q.timeZone;
  return body;
}

File: clients/python/src/pawrly/transports/rest.py
"""``EngineService`` over the JSON REST surface (``pawrly console``)."""

import json
from collections.abc import Iterator

import requests

from ..errors import PawrlyError, UnsupportedByTransport
from ..query import QueryHandle
from ..result import (
    CacheEntryInfo,
    CatalogSnapshot,
    FunctionDescription,
    FunctionInfo,
    HealthReport,
    MaterializeOutcome,
    MaterializeSpec,
    RefreshCatalogOutcome,
    RefreshOutcome,
    ReloadReport,
    SemanticModelDescription,
    SemanticModelInfo,
    SemanticQuery,
    SourceInfo,
    SourceTestReport,
    TableDescription,
    TableInfo,
    VacuumReport,
)
from . import convert


class RestTransport:
    name = "rest"

    def __init__(self, base_url: str, bearer: str | None = None) -> None:
        self._base = base_url.rstrip("/")
        self._session = requests.Session()
        if bearer:
            self._session.headers["authorization"] = f"Bearer {bearer}"

    def _send(
        self,
        method: str,
        path: str,
        json_body: dict | None = None,
        params: dict | None = None,
    ) -> dict:
        resp = self._session.request(
            method, self._base + path, json=json_body, params=params
        )
        body = resp.json() if resp.content else {}
        if not resp.ok:
            err = body.get("error", {}) if isinstance(body, dict) else {}
            raise PawrlyError(
                err.get("code", "PAWRLY_INTERNAL"),
                err.get("message", resp.reason or "request failed"),
            )
        return body

    def query(
        self, sql: str, params: dict[str, str], limit: int | None = None
    ) -> QueryHandle:
        # Stream NDJSON so large results stay memory-bounded. NDJSON carries no id
        # or completion envelope, so `id` is empty and `truncated` stays False.
        payload: dict = {"sql": sql, "params": params, "format": "ndjson"}
        if limit is not None:
            payload["limit"] = limit
        resp = self._session.post(self._base + "/v1/sql", json=payload, stream=True)
        if not resp.ok:
            body = resp.json() if resp.content else {}
            err = body.get("error", {}) if isinstance(body, dict) else {}
            raise PawrlyError(
                err.get("code", "PAWRLY_INTERNAL"),
                err.get("message", resp.reason or "request failed"),
            )
        meta = {"columns": [], "row_count": 0, "truncated": False}
        return QueryHandle("", _ndjson_rows(resp, meta), meta)

    def semantic_query(self, q: SemanticQuery) -> QueryHandle:
        # `/v1/query` returns a buffered JSON envelope (no NDJSON there yet).
        body: dict = {
            "measures": q.measures or [],
            "dimensions": q.dimensions or [],
            "params": q.params or {},
        }
        if q.filters:
            body["filters"] = [
                {"member": f.member, "op": f.op, "values": f.values or []}
                for f in q.filters
            ]
        if q.order_by:
            body["order_by"] = [
                {"member": o.member, "direction": "desc" if o.desc else "asc"}
                for o in q.order_by
            ]
        if q.segments:
            body["segments"] = q.segments
        if q.limit is not None:
            body["limit"] = q.limit
        if q.time_zone:
            body["time_zone"] = q.time_zone
        r = self._send("POST", "/v1/query", body)
        rows = r.get("rows", [])
        meta = {
            "columns": r.get("columns", []),
            "row_count": r.get("row_count", len(rows)),
            "truncated": r.get("truncated", False),
        }
        return QueryHandle("", iter(rows), meta)

    def explain(self, sql: str, analyze: bool) -> str:
        return self._send("POST", "/v1/explain", {"sql": sql, "analyze": analyze}).get(
            "plan", ""
        )

    def cancel(self, query_id: str) -> bool:
        return self._send("POST", f"/v1/queries/{query_id}/cancel").get(
            "cancelled", False
        )

    def materialize(
        self, name: str, spec: MaterializeSpec, namespace: str | None = None
    ) -> MaterializeOutcome:
        body = {"kind": spec.kind, "sql": spec.sql, "params": spec.params or {}}
        r = self._send("PUT", f"/v1/materialized/{name}{_ns_query(namespace)}", body)
        _require_namespace_echo(namespace, r.get("namespace"))
        n = r.get("name", {})
        return MaterializeOutcome(
            name={"schema": n.get("schema", ""), "table": n.get("table", "")},
            file_path=r.get("file_path", ""),
            row_count=r.get("row_count", 0),
            size_bytes=r.get("size_bytes", 0),
        )

    def list_sources(self) -> list[SourceInfo]:
        r = self._send("GET", "/v1/sources")
        return [convert.source_info(s) for s in r.get("sources", [])]

    def list_tables(
        self, source: str | None = None, name_glob: str | None = None
    ) -> list[TableInfo]:
        params = {}
        if source is not None:
            params["source"] = source
        if name_glob is not None:
            params["name_glob"] = name_glob
        r = self._send("GET", "/v1/tables", params=params)
        return [convert.table_info(t) for t in r.get("tables", [])]

    def describe_table(self, name: str) -> TableDescription:
        return convert.table_description(self._send("GET", f"/v1/tables/{name}"))

    def schema_snapshot(
        self, sources: list[str] | None = None, compact: bool = False
    ) -> CatalogSnapshot:
        params = {}
        if sources:
            params["sources"] = ",".join(sources)
        if compact:
            params["compact"] = "true"
        return convert.catalog_snapshot(self._send("GET", "/v1/schema", params=params))

    def cache_entries(self, namespace: str | None = None) -> list[CacheEntryInfo]:
        r = self._send("GET", f"/v1/cache{_ns_query(namespace)}")
        _require_namespace_echo(namespace, r.get("namespace"))
        return [convert.cache_entry(e) for e in r.get("entries", [])]

    def list_functions(self) -> list[FunctionInfo]:
        r = self._send("GET", "/v1/functions")
        return [convert.function_info(f) for f in r.get("functions", [])]

    def describe_function(self, namespace: str, name: str) -> FunctionDescription:
        return convert.function_description(
            self._send("GET", f"/v1/functions/{namespace}/{name}")
        )

    def list_semantic_models(self) -> list[SemanticModelInfo]:
        r = self._send("GET", "/v1/semantic/models")
        return [convert.semantic_model_info(m) for m in r.get("models", [])]

    def describe_semantic_model(self, name: str) -> SemanticModelDescription:
        return convert.semantic_model_description(
            self._send("GET", f"/v1/semantic/models/{name}")
        )

    def list_metrics(self) -> list[dict]:
        return self._send("GET", "/v1/semantic/metrics").get("metrics", [])

    def describe_metric(self, name: str) -> dict:
        return self._send("GET", f"/v1/semantic/metrics/{name}")

    def add_source(self, definition: dict) -> SourceInfo:
        return convert.source_info(self._send("POST", "/v1/sources", definition))

    def remove_source(self, name: str) -> bool:
        return self._send("DELETE", f"/v1/sources/{name}").get("removed", False)

    def test_source(self, name: str) -> SourceTestReport:
        return convert.source_test_report(self._send("POST", f"/v1/sources/{name}/test"))

    def reload_config(self) -> ReloadReport:
        return convert.reload_report(self._send("POST", "/v1/config/reload"))

    def refresh_catalog(self, source: str | None = None) -> RefreshCatalogOutcome:
        params = {"source": source} if source is not None else None
        return convert.refresh_catalog_outcome(
            self._send("POST", "/v1/catalog/refresh", params=params)
        )

    def refresh_table(
        self, name: str, namespace: str | None = None
    ) -> RefreshOutcome:
        r = self._send("POST", f"/v1/tables/{name}/refresh{_ns_query(namespace)}")
        _require_namespace_echo(namespace, r.get("namespace"))
        return convert.refresh_outcome(r)

    def invalidate_cache(self, name: str) -> bool:
        return self._send("DELETE", f"/v1/cache/{name}").get("invalidated", False)

    def vacuum_cache(self) -> VacuumReport:
        return convert.vacuum_report(self._send("POST", "/v1/cache/vacuum"))

    def drop_materialized(self, name: str, namespace: str | None = None) -> bool:
        r = self._send("DELETE", f"/v1/materialized/{name}{_ns_query(namespace)}")
        _require_namespace_echo(namespace, r.get("namespace"))
        return r.get("dropped", False)

    def health(self) -> HealthReport:
        r = self._send("GET", "/v1/health")
        return HealthReport(ok=r.get("ok", False), version=r.get("version", ""))

    def shutdown(self) -> None:
        raise UnsupportedByTransport("shutdown", "rest")

    def close(self) -> None:
        self._session.close()


def _require_namespace_echo(requested: str | None, echoed: str | None) -> None:
    if requested and requested != echoed:
        raise PawrlyError(
            "PAWRLY_PROTOCOL",
            f"server ignored namespace `{requested}` — it predates materialize "
            "namespaces, so the operation targeted the default namespace instead; "
            "upgrade the server",
        )


def _ns_query(namespace: str | None) -> str:
    if not namespace:
        return ""
    return f"?namespace={requests.utils.quote(namespace, safe='')}"


def _ndjson_rows(resp, meta: dict) -> Iterator[dict]:
    try:
        for line in resp.iter_lines():
            if not line:
                continue
            row = json.loads(line)
            if not meta["columns"]:
                meta["columns"] = list(row.keys())
            meta["row_count"] += 1
            yield row
    finally:
        resp.close()
